documentation/advanced/tutorial-i18n-localization.asciidoc

```diff
@@ -169,3 +169,139 @@ public class MyLocale extends Div {
+translated into a RTL Language, such as Arabic. Besides <<Provider sample for
+translation, providing the sample for the Arabic translation>>, at your main
+layout you can add a code like such:
+
+[source,java]
+----
+public class MainLayout extends VerticalLayout {
+
+    public MainLayout() {
+        // ...
+        final UI ui = UI.getCurrent();
+        if (ui.getLocale().getLanguage() == "ar") {
+            ui.setDirection(Direction.RIGHT_TO_LEFT);
+        }
+    }
+}
+----
+
+This will work if changing of locale is only based on the `Accept-Language`
+coming from the client, but if user can define their language, for instance,
+on your application's setting page, then you can make your main layout implement
+the `LocaleChangeObserver` interface, so it will receive the changes of locale
+and then you can set the direction based on the locale set:
+
+[source,java]
+----
+public class MainLayout extends VerticalLayout implements LocaleChangeObserver {
+
+    @Override
+    public void localeChange(LocaleChangeEvent event) {
+        if (event.getLocale().getLanguage() == "ar") {
+            event.getUI().setDirection(Direction.RIGHT_TO_LEFT);
+        } else {
+            event.getUI().setDirection(Direction.LEFT_TO_RIGHT);
+        }
+    }
+}
+----
+
+==== Frontend projects
+
+For frontend applications only, to set the RTL mode you can call `document.dir = 'rtl'`.
+
+==== Adding RTL support to your custom elements
+
+If you have your own custom elements, there are a few steps needed in order to
+add RTL support to them:
+
+. If your element extends Vaadin's `ElementMixin`, make sure that it depends on
+the version starting from `2.3.0`. If not, you can make the element extends it
+or `DirMixin` only (`DirMixin` is part of the `@vaadin/vaadin-element-mixin`
+package).
+
++
+[source,javascript]
+----
+import { PolymerElement } from '@polymer/polymer/polymer-element.js';
+import { DirMixin } from '@vaadin/vaadin-element-mixin/vaadin-dir-mixin.js';
+
+class MyElement extends DirMixin(PolymerElement) {}
+----
+
++
+The `DirMixin` registers the element to changes on the `dir` attribute at the
+document level and keeps it in sync with the element's `dir` attribute. That's
+helpful to easily check the RTL status in both CSS and JS code.
+
+. Make sure your styles are adjusted properly for the RTL mode.
++
+For example, if you define values for the padding on the `:host`, like:
+
++
+[source,css]
+----
+:host {
+    padding-right: 1em;
+    padding-left: 2em;
+}
+----
++
+You may want to define the proper style for the RTL, like:
++
+[source,css]
+----
+:host([dir="rtl"]) {
+    padding-right: 2em;
+    padding-left: 1em;
+}
+----
++
+You may want to pay attention to declarations such as `padding`, `margin`,
+`text-align`, `float` and `transform` on your styles. Flex containers usually
+are handled well by the browser and don't require extra work. More details can
+https://rtlstyling.com/posts/rtl-styling/[be found here].
```

Comment:
Generally, avoid “click here” type of link text. Would be better to provide a descriptive title of the page that it links to.

Reply:
Done.